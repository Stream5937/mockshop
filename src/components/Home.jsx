import { useOutletContext, Link, useParams } from "react-router-dom";
import style from "../styles/home.module.css";

export default function Home() {
  const context = useOutletContext();
  console.log("context: ", context);
  const { products, error, loading } = context[0];
  console.log("products", products);

  //defines a new array from the set of unique categories
  const categories = [...new Set(products.map((p) => p.category))];

  let copyCategories = [...categories];

  let initialProducts = products.map((p) => {
    if (copyCategories.includes(p.category)) {
      //remove category from copy arrray
      const index = copyCategories.indexOf(p.category);
      if (index > -1) {
        // only splice array when item is found
        copyCategories.splice(index, 1); // 2nd parameter means remove one item only
      }
      //return the product
      return p;
    }
  });

  initialProducts = initialProducts.filter((prod) => prod !== undefined);

  console.log("initial products: ", initialProducts);

  return (
    <>
      <h2>MOCK SHOP THIS WEEKS PROMOTIONS</h2>
      <div className={style.homeContainer}>
        {initialProducts.map((product) => (
          <div id={style.product} key={product.id}>
            <img id={style.imgHome} src={product.image} alt={product.title} />
            <h3>{product.title}</h3>

            <p>{product.description.substring(0, 80)}...</p>

            <Link to={`/product/${product.id}`}>View Product</Link>
          </div>
        ))}
      </div>
    </>
  );
}
