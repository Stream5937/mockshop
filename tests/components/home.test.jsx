import { describe, it, expect, vi, beforeEach } from "vitest";
import { renderWithRouter } from "../utils/renderWithRouter.jsx";
import { screen } from "@testing-library/react";
import Home from "../../src/components/Home.jsx";

vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual("react-router-dom");

  const mockProduct = {
    id: 1,
    title: "Test Product",
    category: "Test Category",
    price: 10,
    description: "This is a test product description",
    rating: { rate: 4.5, count: 20 },
    image: "test.jpg",
  };

  const mockContext = [
    { products: [mockProduct], error: null, loading: false }, // context[0]
    [[{ item: mockProduct, number: 1 }], vi.fn()], // context[1]
    [vi.fn(), vi.fn(), vi.fn()], // context[2]
    [1, vi.fn()], // context[3]
  ];

  return {
    ...mod,
    useOutletContext: () => mockContext, // return array
  };
});

describe("Home component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders with the categories of products", () => {
    renderWithRouter(<Home />);
    // Check snapshot
    expect().toMatchSnapshot();
  });

  // Check important UI bits
  it("renders correct text in h3 elements", () => {
    renderWithRouter(<Home />);
    const heading = screen.getByRole("heading", { level: 3 });
    expect(heading).toHaveTextContent("Test Product");
  });
});
