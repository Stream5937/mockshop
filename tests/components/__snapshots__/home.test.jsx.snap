// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home component > renders with the categories of products 1`] = `undefined`;
